Readme.md

# express-trace

  Express tracer and middleware profiler.

## Installation

     $ npm install express-trace

__NOTE__: supports express >= 2.3.5

## Example

 Suppose we have the following express app with a few routes that can satisfy a url with the form `GET /user/NAME`, and some middlware

    app.use(express.favicon());
    app.use(express.bodyParser());
    app.use(express.methodOverride());
    app.use(express.cookieParser());
    app.use(express.session({ secret: 'foo' }));
    app.use(express.static(__dirname));

    app.get('/user/:name', function(req, res, next){
      setTimeout(next, 1000);
    });

    app.get('/user/:name', function(req, res, next){
      setTimeout(next, 200);
    });

    app.get('/user/tobi', function(req, res, next){
      res.send('loaded tobi');
    });

all we need to do before we `listen()`, is `require()` express-trace and apply it to the app. This essentially wraps all of the middleware and routes with functions used for reporting.

    require('express-trace')(app);

stderr for `GET /user/tobi`:

    GET /user/tobi
    middleware / anonymous 0ms
    middleware / favicon 0ms
    middleware / bodyParser 0ms
    middleware / methodOverride 0ms
    middleware / cookieParser 0ms
    middleware / session 0ms
    middleware / static 1ms
    middleware / router
      app.get(/user/:name) 1000ms
      app.get(/user/:name) 200ms
      app.get(/user/tobi) 
    responded to GET /user/tobi in 1205ms with 200 "OK"

## License 

(The MIT License)

Copyright (c) 2011 TJ Holowaychuk &lt;tj@vision-media.ca&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.0.2 / 2011-05-20 
==================

  * Added route middleware support. Closes #1

0.0.1 / 2011-05-19 
==================

  * Initial release


lib/express-trace.js

/*!
 * express-trace
 * Copyright(c) 2011 TJ Holowaychuk <tj@vision-media.ca>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var http = require('http');

/**
 * Export trace function.
 */

exports = module.exports = trace;

/**
 * Library version.
 */

exports.version = '0.0.2';

/**
 * Status code color map.
 */

var colors = {
    2: 32
  , 3: 36
  , 4: 33
  , 5: 31
};

/**
 * Trace middleware in the given `app`.
 *
 * @param {express.HTTPServer} app
 * @api public
 */

function trace(app) {
  var stack = app.stack
    , len = stack.length;

  for (var i = 0; i < len; ++i) {
    stack[i].handle = (function(route, fn){
      // TODO: mounted apps
      // TODO: route middleware

      // router
      if ('router' == fn.name) {
        for (var method in fn.routes) {
          fn.routes[method].forEach(function(route){
            var callback = route.callback;

            // callback
            route.callback = function(req, res, next){
              var start = start = new Date;;
              
              process.stderr.write('    \033[90mapp.'
                + route.method.toLowerCase() + '('
                + route.path + ')\033[0m ');

              callback(req, res, function(err){
                console.error('\033[90m%dms\033[0m', new Date - start);
                next(err);
              });
            };

            // middleware
            route.middleware = route.middleware.map(function(fn){
              return function(req, res, next){
                var name = fn.name || 'anonymous'
                  , start = new Date;

                process.stderr.write('      \033[90mmiddleware '
                  + '\033[36m' + name + '\033[0m ');

                fn(req, res, function(err){
                  console.error('\033[90m%dms\033[0m', new Date - start);
                  next(err);
                });
              }
            });
          });
        }
      }

      // regular middleware
      return function(req, res, next){
        var route = route || '/'
          , name = fn.name || 'anonymous'
          , router = 'router' == fn.name
          , start = new Date;

        // middleware
        process.stderr.write('  \033[90mmiddleware \033[33m'
          + route + ' \033[36m'
          + name + '\033[0m'
          + (router ? '\n' : ' '));

        // duration
        fn(req, res, function(err){
          console.error((router ? '  ' : '')
            + '\033[90m%dms\033[0m', new Date - start);
          next(err);
        });
      }
    })(stack[i].route, stack[i].handle);
  }

  stack.unshift({
      route: ''
    , handle: function(req, res, next){
      var start = new Date;
      res.on('finish', function(){
        var color = colors[res.statusCode / 100 | 0];
        console.error('\n  \033[90mresponded to %s \033[33m%s\033[0m '
          + '\033[90min %dms with \033[' + color + 'm%s\033[0m'
          + ' \033[90m"%s"\033[0m'
          , req.method
          , req.url
          , new Date - start
          , res.statusCode
          , http.STATUS_CODES[res.statusCode]);
      });
      console.error('\n  \033[90m%s \033[33m%s\033[0m', req.method, req.url);
      next();
    }
  });

  stack.push({
      route: ''
    , handle: function(req, res, next){
      next();
    }
  });
  
  stack.push({
      route: ''
    , handle: function(err, req, res, next){
      req.__err = err;
      next(err);
    }
  });
};

examples/basic.js

/**
 * Module dependencies.
 */

var trace = require('../')
  , express = require('express')
  , app = express.createServer();

app.use(express.favicon());
app.use(express.bodyParser());
app.use(express.methodOverride());
app.use(express.cookieParser());
app.use(express.session({ secret: 'foo' }));
app.use(express.static(__dirname));

function ensureUsername(name) {
  return function ensureUsername(req, res, next){
    setTimeout(function(){
      if (req.params.name == name) {
        next();
      } else {
        next('route');
      }
    }, Math.random() * 300);
  }
}

function rejectName(name) {
  return function rejectName(req, res, next){
    if (name == req.params.name) return res.send(401);
    next();
  }
}

app.get('/user/:name', rejectName('loki'));

app.get('/user/:name', ensureUsername('tj'), function(req, res, next){
  res.send('loaded tj');
});

app.get('/user/:name', function(req, res, next){
  setTimeout(next, 1000);
});

app.get('/user/:name', function(req, res, next){
  setTimeout(next, 200);
});

app.get('/user/tobi', function(req, res, next){
  res.send('loaded tobi');
});

app.get('/file/*', function(req, res, next){
  var file = req.params[0];
  res.send('requested ' + file);
});

app.get('/error', function(req, res, next){
  asdf();
});

trace(app);

app.listen(3000);